File: src/fr/gstraymond/ui/FacetListAdapter.java
package fr.gstraymond.ui;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import android.content.Context;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.ExpandableListView;
import android.widget.TextView;
import fr.gstraymond.R;
import fr.gstraymond.biz.SearchOptions;
import fr.gstraymond.constants.FacetConst;
import fr.gstraymond.magicsearch.model.response.facet.Facet;
import fr.gstraymond.magicsearch.model.response.facet.Term;


public class FacetListAdapter extends BaseExpandableListAdapter {

	private List<String> facetList;
	private List<String> selectedFacets;
	private Map<String, Facet> facetMap;
	private List<Term> selectedTerms;

	public FacetListAdapter(Map<String, Facet> facets, SearchOptions options) {
		this.facetMap = facets;
		this.selectedFacets = new ArrayList<String>();
		this.facetList = new ArrayList<String>();
		this.selectedTerms = new ArrayList<Term>();

		if (facets == null) {
			return;
		}
		
		for (String facetAsString : FacetConst.getFacetOrder()) {
			Facet facet = facets.get(facetAsString);
			if (facet == null) {
				continue;
			}
			facetList.add(facetAsString); 
			List<Term> facetTerms = facet.getTerms();
			
			if (! facetTerms.isEmpty()) {
				List<String> termsAsString = options.getFacets().get(facetAsString);
				if (termsAsString != null) {
					selectedFacets.add(facetAsString);
					selectedTerms.addAll(findTerms(termsAsString, facetTerms));
				}
			}
		}
	}
	
	private List<Term> findTerms(List<String> termsAsString, List<Term> terms) {
		List<Term> termsFound = new ArrayList<Term>();
		
		for (String termAsString : termsAsString) {
			Term termFound = findTerm(termAsString, terms);
			if (termFound != null) {
				termsFound.add(termFound);
			}
		}
		
		return termsFound;
	}
	
	private Term findTerm(String termAsString, List<Term> terms) {
		for(Term term : terms) {
			if (term.getTerm().equals(termAsString)) {
				return term;
			}
		}
		return null;
	}

	private List<Term> getChildren(int groupPosition) {
		return facetMap.get(getGroup(groupPosition)).getTerms();
	}

	@Override
	public Object getChild(int groupPosition, int childPosition) {
		return getChildren(groupPosition, childPosition);
	}

	@Override
	public long getChildId(int groupPosition, int childPosition) {
		return childPosition;
	}

	@Override
	public View getChildView(int groupPosition, int childPosition, boolean isLastChild, View convertView,
			ViewGroup parent) {
		View view = convertView;
        if (view == null) {
            LayoutInflater inflater = getLayoutInflater(parent.getContext());
            view = inflater.inflate(R.layout.drawer_child, null);
        }
        
		Term term = getChildren(groupPosition, childPosition);
		String text = term.getTerm();

		TextView textTextView = (TextView) view.findViewById(R.id.drawer_child_text);
		TextView counterTextViewInactive = (TextView) view.findViewById(R.id.drawer_child_counter_inactive);
		counterTextViewInactive.setVisibility(View.GONE);
		TextView counterTextViewActive = (TextView) view.findViewById(R.id.drawer_child_counter_active);
		counterTextViewActive.setVisibility(View.GONE);
		
		TextView counterTextView = counterTextViewInactive;
		
		if (selectedTerms.contains(term)) {
			counterTextView = counterTextViewActive;
			textTextView.setText(Html.fromHtml("<b>" + text + "</b>", null, null));
		} else {
			textTextView.setText(text);
		}
		counterTextView.setVisibility(View.VISIBLE);
		counterTextView.setText(term.getCount() + "");
		return view;
	}

	private Term getChildren(int groupPosition, int childPosition) {
		return getChildren(groupPosition).get(childPosition);
	}

	@Override
	public int getChildrenCount(int groupPosition) {
		return getChildren(groupPosition).size();
	}

	@Override
	public Object getGroup(int groupPosition) {
		return facetList.get(groupPosition);
	}

	@Override
	public int getGroupCount() {
		return facetList.size();
	}

	@Override
	public long getGroupId(int groupPosition) {
		return groupPosition;
	}

	@Override
	public View getGroupView(int groupPosition, boolean isExpanded, View convertView, ViewGroup parent) {
		View view = convertView;
        if (view == null) {
            LayoutInflater inflater = getLayoutInflater(parent.getContext());
            view = inflater.inflate(R.layout.drawer_group, null);
        }
		
        TextView textView = (TextView) view.findViewById(R.id.drawer_group_textview);
		String facet = facetList.get(groupPosition);
		
		if (selectedFacets.contains(facet)) {
			ExpandableListView expandableListView = (ExpandableListView) parent;
		    expandableListView.expandGroup(groupPosition);
		}
		
		textView.setText(FacetConst.getFacetName(facet, parent.getContext()));
		return view;
	}

	@Override
	public boolean hasStableIds() {
		return false;
	}

	@Override
	public boolean isChildSelectable(int groupPosition, int childPosition) {
		return true;
	}

	public String getFacet(Term term) {
		for (Map.Entry<String, Facet> facet : facetMap.entrySet()) {
			if (facet.getValue().getTerms().contains(term)) {
				return facet.getKey();
			}
		}
		return null;
	}
	
	public boolean isTermSelected(Term term) {
		return selectedTerms.contains(term);
	}
	
	private LayoutInflater getLayoutInflater(Context context) {
		return LayoutInflater.from(context);
	}
}

File: app/src/main/kotlin/fr/gstraymond/ui/adapter/FacetListAdapter.kt
package fr.gstraymond.ui.adapter

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.BaseExpandableListAdapter
import android.widget.TextView
import fr.gstraymond.R
import fr.gstraymond.biz.SearchOptions
import fr.gstraymond.constants.FacetConst
import fr.gstraymond.models.search.response.Aggregations
import fr.gstraymond.models.search.response.Bucket
import fr.gstraymond.utils.find
import fr.gstraymond.utils.gone
import fr.gstraymond.utils.inflate
import fr.gstraymond.utils.visible


class FacetListAdapter(aggregationsMap: Map<String, Aggregations>?,
                       private val options: SearchOptions, context: Context) : BaseExpandableListAdapter() {
    private val mutableFacetMap = aggregationsMap?.toMutableMap() ?: mutableMapOf()
    private val facetList = mutableListOf<String>()
    private val selectedFacets = mutableListOf<String>()
    private val selectedTerms = mutableListOf<Bucket>()

    init {
        for (facetAsString in FacetConst.getFacetOrder()) {
            val facet = mutableFacetMap[facetAsString]

            if (facet == null || facet.buckets.isEmpty()) {
                mutableFacetMap.remove(facetAsString)
                continue
            }

            facetList.add(facetAsString)
            val facetTerms = facet.buckets

            if (!facetTerms.isEmpty()) {
                val termsAsString = options.facets[facetAsString]
                if (termsAsString != null) {
                    selectedFacets.add(facetAsString)
                    selectedTerms.addAll(findTerms(termsAsString, facetTerms))
                }
            }
        }

        for ((facetAsString, facet) in mutableFacetMap) {
            if (showLoadMore(facet, facetAsString)) {
                val loadMoreTerm = Bucket(context.getString(R.string.facet_more), -1)
                facet.buckets.add(loadMoreTerm)
            }
        }
    }

    private fun showLoadMore(aggregations: Aggregations, facetAsString: String): Boolean {
        val facetSizeRequested = options.facetSize[facetAsString] ?: 10
        return aggregations.buckets.size == facetSizeRequested
    }

    private fun findTerms(termsAsString: List<String>, buckets: List<Bucket>) = termsAsString.mapNotNull { findTerm(it, buckets) }

    private fun findTerm(termAsString: String, buckets: List<Bucket>) = buckets.firstOrNull { it.key == termAsString }

    private fun getChildren(groupPosition: Int): List<Bucket> = mutableFacetMap[getGroup(groupPosition)]!!.buckets

    override fun getChild(groupPosition: Int, childPosition: Int) = getTerm(groupPosition, childPosition)

    override fun getChildId(groupPosition: Int, childPosition: Int) = childPosition.toLong()

    override fun getChildView(groupPosition: Int,
                              childPosition: Int,
                              isLastChild: Boolean,
                              convertView: View?,
                              parent: ViewGroup): View {
        val view = convertView ?: parent.context.inflate(R.layout.drawer_child)

        val term = getTerm(groupPosition, childPosition)

        view.find<TextView>(R.id.drawer_child_text).text = term.key

        val counterTextViewInactive = view.find<TextView>(R.id.drawer_child_counter_inactive)
        val counterTextViewActive = view.find<TextView>(R.id.drawer_child_counter_active)

        val (counterTextView, hiddenCounterTextView) = if (selectedTerms.contains(term)) {
            counterTextViewActive to counterTextViewInactive
        } else {
            counterTextViewInactive to counterTextViewActive
        }

        hiddenCounterTextView.gone()

        counterTextView.visible()
        counterTextView.text =
                if (term.doc_count > 0) term.doc_count.toString() + ""
                else "?"

        return view
    }

    fun getTerm(groupPosition: Int, childPosition: Int): Bucket = getChildren(groupPosition)[childPosition]

    override fun getChildrenCount(groupPosition: Int): Int = getChildren(groupPosition).size

    override fun getGroup(groupPosition: Int) = getFacet(groupPosition)

    fun getFacet(groupPosition: Int): String = facetList[groupPosition]

    override fun getGroupCount() = facetList.size

    override fun getGroupId(groupPosition: Int) = groupPosition.toLong()

    override fun getGroupView(groupPosition: Int, isExpanded: Boolean, convertView: View?, parent: ViewGroup): View {
        val view = convertView ?: parent.context.inflate(R.layout.drawer_group)

        val facet = facetList[groupPosition]
        val groupTextView = view.find<TextView>(R.id.drawer_group_textview)
        groupTextView.text = FacetConst.getFacetName(facet, parent.context)

        val selectedTextView = view.find<TextView>(R.id.drawer_group_selected_textview)
        if (selectedFacets.contains(facet)) {
            val children = getChildren(groupPosition).filter { selectedTerms.contains(it) }
            selectedTextView.text = children.map(Bucket::key).joinToString()
            selectedTextView.visible()
        } else {
            selectedTextView.gone()
        }

        return view
    }

    override fun hasStableIds() = false

    override fun isChildSelectable(groupPosition: Int, childPosition: Int) = true

    fun isTermSelected(bucket: Bucket) = selectedTerms.contains(bucket)
}
